app/src/main/java/com/randomappsinc/studentpicker/backupandrestore/backup/BackupDataFragment.kt

```diff
@@ -0,0 +1,149 @@
+    }
+
+    fun setBackupSubtitle() {
+        if (BackupDataManager.getBackupUri(context) == null) {
+            binding.backupSubtitle.setText(R.string.backup_data_explanation)
+        } else {
+            val lastBackupUnixTime: Long = preferencesManager.lastBackupTime
+            val lastBackupTime: String = TimeUtils.getLastBackupTime(lastBackupUnixTime)
+            binding.backupSubtitle.text = requireContext().getString(
+                    R.string.backup_subtitle_with_backup,
+                    lastBackupTime)
+        }
+    }
+
+    private fun backupDataClicked() {
+        binding.backupData.setOnClickListener {
+            if (PermissionUtils.isPermissionGranted(Manifest.permission.WRITE_EXTERNAL_STORAGE, context)) {
+                backupData()
+            } else {
+                PermissionUtils.requestPermission(
+                        this,
+                        Manifest.permission.WRITE_EXTERNAL_STORAGE,
+                        WRITE_EXTERNAL_STORAGE_CODE)
+            }
+        }
+    }
+
+    private fun backupData() {
+        val backupUri: String? = BackupDataManager.getBackupUri(context)
+        if (backupUri == null) {
+            chooseBackupLocation()
+        } else {
+            BackupDataManager.backupData(requireContext(), true)
+        }
+    }
+
+    private fun chooseBackupLocation() {
+        UIUtils.showLongToast(R.string.choose_backup_folder, context)
+        val intent = Intent(Intent.ACTION_CREATE_DOCUMENT)
+        intent.addCategory(Intent.CATEGORY_OPENABLE)
+        intent.type = "text/plain"
+        intent.putExtra(Intent.EXTRA_TITLE, BackupDataManager.BACKUP_FILE_NAME)
+        startActivityForResult(intent, WRITE_BACKUP_FILE_CODE)
+    }
+
+   private fun changeBackupFolder() {
+        binding.changeBackupFolder.setOnClickListener {
+            if (PermissionUtils.isPermissionGranted(Manifest.permission.WRITE_EXTERNAL_STORAGE, context)) {
+                chooseBackupLocation()
+            } else {
+                PermissionUtils.requestPermission(
+                        this,
+                        Manifest.permission.WRITE_EXTERNAL_STORAGE,
+                        WRITE_EXTERNAL_STORAGE_CODE)
+            }
+        }
+    }
+
+    private val backupDataListener: BackupDataManager.Listener = object : BackupDataManager.Listener {
+        override fun onBackupComplete() {
+            UIUtils.showShortToast(R.string.backup_successful, context)
+            setBackupSubtitle()
+        }
+
+        override fun onBackupFailed() {
+            UIUtils.showLongToast(R.string.backup_failed, context)
+        }
+    }
+
+    override fun onActivityResult(requestCode: Int, resultCode: Int, resultData: Intent?) {
+        if (requestCode == WRITE_BACKUP_FILE_CODE && resultCode == Activity.RESULT_OK && resultData != null) {
+            val context = context
+            val uri = resultData.data
+            if (uri == null || context == null) {
```

Comment:
I don't think we should null check context here (feels like unnecessary error suppression). I think we should just do requireContext() in the block where we need it.